oslo_vmware/image_transfer.py

```diff
@@ -47,19 +50,49 @@ def _create_progress_updater(handle):
         return updater
 
 
-def _start_transfer(read_handle, write_handle, timeout_secs):
+def _start_transfer(read_handle, write_handle, timeout_secs, queue_size=64):
     # read_handle/write_handle could be an NFC lease, so we need to
     # periodically update its progress
     read_updater = _create_progress_updater(read_handle)
     write_updater = _create_progress_updater(write_handle)
 
     timer = timeout.Timeout(timeout_secs)
-    try:
+    read_event = event.Event()
+    write_event = event.Event()
+    data_queue = queue.LightQueue(maxsize=queue_size)
+
+    def _inner_write():
         while True:
-            data = read_handle.read(CHUNK_SIZE)
-            if not data:
+            try:
+                if read_event.ready():
+                    write_handle.write(data_queue.get_nowait())
+                else:
+                    write_handle.write(data_queue.get(True, timeout_secs))
+            except queue.Empty:
+                if read_event.ready():
+                    write_event.send()
+                    break
+            except Exception as e:
+                write_event.send_exception(e)
                 break
-            write_handle.write(data)
+
+    def _inner_read():
+        while True:
+            try:
+                data = read_handle.read(CHUNK_SIZE)
+                if not data:
+                    read_event.send()
+                    break
+                data_queue.put(data, True, timeout_secs)
```

Comment:
Yes, doesn't make much sense, since after each iteration `timeout_secs` remains the same.
But, anyway, If we'll kill the thread on timeout, then we don't need to pass a timeout to `put` or `get` anymore.